src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -1565,29 +1566,27 @@ public void Dispose()
 			using (new SessionIdLoggingContext(SessionId))
 			{
 				log.Debug(string.Format("[session-id={0}] running ISession.Dispose()", SessionId));
+				TransactionContext?.Wait();
 				var transactionContext = TransactionContext;
-				if (transactionContext != null)
+				if (transactionContext != null && transactionContext.CanFlushOnSystemTransactionCompleted)
 				{
-					transactionContext.Wait();
-					if (transactionContext.CanFlushOnSystemTransactionCompleted)
+					// absolute nogo to remove before merge, done for illustrating a race condition
+					try
 					{
-						// absolute nogo to remove before merge, done for illustrating a race condition
-						try
-						{
-							// Bloat only distributed case
-							if (((Transaction.AdoNetWithSystemTransactionFactory.SystemTransactionContext)transactionContext)
-							    .AmbientTransaction.TransactionInformation.DistributedIdentifier != Guid.Empty)
-								System.Threading.Thread.Sleep(500);
-						}
-						catch
-						{
-							// do not care
-						}
-						// end absolute nogo
-						transactionContext.ShouldCloseSessionOnSystemTransactionCompleted = true;
-						return;
+						// Bloat only distributed case
+						if (((Transaction.AdoNetWithSystemTransactionFactory.SystemTransactionContext) transactionContext)
```

Comment:
That one was occurring because the synchronization point was broken by the commented `if`! If we should guards against maybe broken synchronization code, we will end up coding over protected code everywhere.

Granted, here it is easy to guard against a synchronization failure causing the thread context to still get changed after the `Wait`. But of course the local variable must be set after the `Wait` then checked. Still I am not keen on supposing the synchronization is broken for coding afterwar trying to "protect" against it being broken: it could hide that very fact, synchronization being broken, and cause it to be harder to diagnose. As always, better fails early. So yes, I rather have that null ref occuring if the synchronization ends up broken, as it is currently by the lack of that commented out `if`.